momo_modules/multi_sensor_module/src/pulse.c

```diff
@@ -18,39 +38,162 @@ void pulse_sample()
 	set_analog_power(1);
 
 	/*
-	 * Count pulses for 0.1 seconds
+	 * Count pulses for 500 ms recording the number
+	 * of complete pulses (falling edge, low period, rising edge)
 	 */
 
-	++periods;
-
-	tmr_config(2, kTMRPrescale_64, kTMRPostscale1_16);
-	tmr_loadperiod(2, 98);
-	tmr_load(2, 0);
-	tmr_flag(2) = 0;
-	tmr_setstate(2, 1);
+	pulses = 0;
+	invalid_pulses = 0;
+	pulse_start = 0;
+	last_valid_start = 0;
 	
+	//Use 16 bit timer to capture pulse widths and intervals
+	tmr1_config(pack_tmr1_config(kInstructionClock, kPrescaler1_8));
+	tmr1_load(kTimer1_HalfSecond);
+	tmr1_setstate(kEnabled_Sync);
+
 	//Wait for the pin to go high so that we have a valid signal
-	while(PIN(PULSE_IN) == 0 && tmr_flag(2) == 0)
+	while(PIN(PULSE_IN) == 0 && tmr1_flag == 0)
 		;
 	
-	if (tmr_flag(2) == 0)
+	if (tmr1_flag == 0)
 	{
 		ioc_flag_b(PULSE_IOC) = 0;
 		ioc_detect_falling_b(PULSE_IOC, 1);
+		ioc_detect_rising_b(PULSE_IOC, 1);
 
-		while(tmr_flag(2) == 0)
+		//Actual counting happens in interrupt code
+		while(tmr1_flag == 0)
 			;
 
 		//Disable everything
 		ioc_detect_falling_b(PULSE_IOC, 0);
+		ioc_detect_rising_b(PULSE_IOC, 0);
 	}
 
-	tmr_setstate(2, 0);
+	tmr1_setstate(kDisabled);
 	set_analog_power(0);
 	PIN_DIR(PULSE_IN, OUTPUT);
 }
 
+void pulse_falling_edge()
+{
+	if (pulses < MAX_PULSES)
```

Comment:
I think with the logic in pulse_rising_edge this will confuse things when pulses >= MAX_PULSES.  It's probably harmless, but it's weird behavior.  This check should be removed or everything in pulse_rising_edge should be guarded.